keps/sig-node/2000-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,864 @@
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+*   As a cluster administrator, I can configure the nodes in my cluster to
+    allocate X seconds for my pods to terminate gracefully during a node
+    shutdown
+
+#### Story 2
+
+*   As a developer I can expect that my pods will terminate gracefully during
+    node shutdowns
+
+### Background on Linux Shutdown
+
+In the context of this KEP, shutdown is referred to as shutdown of the
+underlying machine. On most linux distros shutdown can be initiated via a
+variety of methods for example:
+
+1. `shutdown -h now`
+2. `shutdown -h +30` `#schedule a delayed shutdown in 30mins`
+3. `systemctl poweroff`
+4. Physically pressing the power button on the machine
+5. If a machine is a VM, the underlying hypervisor can press the “virtual”
+   power button
+6. For a cloud instance, stopping the instance via Cloud API, e.g. via `gcloud
+   compute instances stop`. Depending on the cloud provider, this may result in
+   virtual power button press by the underlying hypervisor.
+
+
+Some of these cases will involve the machine receiving an ACPI event to change
+the power state. The machine can go from G0 (working state) to G2 (Soft Off)
+and finally to G3 (Off) [more info on ACPI
+states](https://en.wikipedia.org/wiki/Advanced_Configuration_and_Power_Interface).
+On Linux, prior to shutdown usually a system demon will listen to these events
+and perform some series of actions prior to userspace calling the
+[reboot(2)](https://man7.org/linux/man-pages/man2/reboot.2.html) systemcall
+with `LINUX_REBOOT_CMD_POWER_OFF` or `LINUX_REBOOT_CMD_HALT` to actually
+shutdown the machine.
+
+Historically, ACPI events were often handled by the
+[acpid](https://wiki.archlinux.org/index.php/Acpid) daemon which uses a variety
+of mechanisms to watch ACPI events (i.e. reading `/proc/acpi/event` or
+`/dev/input/eventX` to react to power button presses). However, in most modern
+linux distros today,
+[systemd-logind](https://www.freedesktop.org/wiki/Software/systemd/logind/)
+has taken over as the main component [reacting to ACPI
+events](https://wiki.archlinux.org/index.php/Power_management#ACPI_events) and
+initiating shutdown of the machine. On a system with
+systemd-logind, for example, a trigger of the power button will
+result in the systemd target
+[poweroff](https://www.freedesktop.org/software/systemd/man/systemd-halt.service.html)
+being run (see
+[HandlePowerKey](https://www.freedesktop.org/software/systemd/man/logind.conf.html),
+which will terminate all the systemd services running on the machine and
+eventually shut it down. However, in the context of kubernetes, systemd is not
+aware of the pods and containers running on the machine and systemd will simply
+kill them as regular linux processes.
+
+### Background on Inhibitors
```

Comment:
Windows has similar functionality to inhibitor locks.
https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-shutdownblockreasoncreate?redirectedfrom=MSDN is the API call to delay a shutdown.

Can we make the specifics of how kubelet controls the shutdown delay configurable for different platforms?

Reply:
I can create a Background on Windows Shutdown section for inclusion in the KEP if that would be useful.

Reply:
Thanks @marosset, it would be great to add a Windows specific section, thanks for offering your help!

I'm thinking it would be good to merge the initial version of the KEP and then you can do a followup PR to add the windows specific info, WDYT?

Reply:
Sure, I can make a followup PR with the windows specific issue.

Reply:
Thanks Mark, will ping you after to get the windows specific info added.